### PR TITLE
8275317: AArch64: Support some type conversion vectorization in SLP

### PR DESCRIPTION
**Pull request by fg1417**
Current SLP vectorizer in C2 compiler doesn't support type conversion
operations. But AArch64 has vector type conversion instructions in
both NEON and SVE.

The type conversion involves two kinds of scenarios, conversion between
the same data sizes and conversion between different data sizes. If we
want to support casts between different data sizes, we need to amend
the code part for identifying adjacent memory references and the code
part for justifying if the combination is profitable. I suppose it
would be easier to review if we split the whole task to support type
conversion into two separate patches, one for the same data sizes and
the other one for different data sizes. The goal of this patch is just
to support conversions within the same data size, including:
  int -> float
  float -> int
  long -> double
  double -> long

A typical test case:

for (int i = start; i < limit; i++) {
    b[i] = (float) a[i];
}

To implement it, the patch completed the necessary instructions and
matching rules in the backend and added implementation for SLP in
the middle end.

The percentage of performance uplift on aarch64 system:
Mode: avgt
Cnt: 15
Metric: (ns/op)

benchmark                       percentage change  [(After-Before)/Before] 
VectorLoop.convertD2L      -48.46%	                 
VectorLoop.convertF2I        -55.67%
VectorLoop.convertI2F        -55.27%
VectorLoop.convertL2D       -48.75%

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275317](https://bugs.openjdk.java.net/browse/JDK-8275317): AArch64: Support some type conversion vectorization in SLP

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Nick Gasson](https://openjdk.java.net/census#ngasson) (@nick-arm - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6145/head:pull/6145` \
`$ git checkout pull/6145`

Update a local copy of the PR: \
`$ git checkout pull/6145` \
`$ git pull https://git.openjdk.java.net/jdk pull/6145/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6145`

View PR using the GUI difftool: \
`$ git pr show -t 6145`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6145.diff">https://git.openjdk.java.net/jdk/pull/6145.diff</a>

</details>

### CODE REVIEW

**Comment by fg1417** [comment]
Hi, the patch mostly benefits AArch64, but the changes are all shared code. Could anyone help review it ?

**Review by TobiHartmann — COMMENTED** [review]
That looks good to me but x86 supports vector instructions for these operations as well, right? Or am I missing something?

https://github.com/openjdk/jdk/blob/55b36c6f3bb7eb066daaf41f9eba46633afedf08/src/hotspot/cpu/x86/x86.ad#L6701

Do you have perf numbers for x86?

**Comment by fg1417** [comment]
> That looks good to me but x86 supports vector instructions for these operations as well, right? Or am I missing something?
> 
> https://github.com/openjdk/jdk/blob/55b36c6f3bb7eb066daaf41f9eba46633afedf08/src/hotspot/cpu/x86/x86.ad#L6701
> 
> Do you have perf numbers for x86?

Yes, I got perf data on X86 as showed below:

Before the patch:
  Benchmark              (length)  Mode  Cnt    Score   Error  Units
  VectorLoop.convertD2L       523  avgt   15  527.330 ± 6.159  ns/op
  VectorLoop.convertF2I       523  avgt   15  545.808 ± 4.677  ns/op
  VectorLoop.convertI2F       523  avgt   15  373.227 ± 1.259  ns/op
  VectorLoop.convertL2D       523  avgt   15  869.646 ± 0.183  ns/op

After the patch:
  Benchmark              (length)  Mode  Cnt    Score   Error  Units
  VectorLoop.convertD2L       523  avgt   15  530.785 ± 4.767  ns/op
  VectorLoop.convertF2I       523  avgt   15  545.831 ± 7.576  ns/op
  VectorLoop.convertI2F       523  avgt   15   66.562 ± 2.270  ns/op
  VectorLoop.convertL2D       523  avgt   15  869.510 ± 0.075  ns/op

X86 supports int to FP only, and got performance uplift on convertI2F.  But it has implementation limitation on both FP to integer types and double to long, and got no obvious positive effect on these scenarios.

**Review by TobiHartmann — APPROVED** [review]
Thanks for sharing these numbers. Your changes look good to me and our internal testing all passed.

**Comment by TobiHartmann** [comment]
A second review would be good.

/reviewers 2

**Comment by fg1417** [comment]
> A second review would be good.
> 
> /reviewers 2

Thanks for your review, @TobiHartmann . Can I get a second review please?

**Comment by jatin-bhateja** [comment]
Hi @fg1417 , Thanks for fixing this, I have X86 backend changes for missing same sized vector cast operations. Will push for review after this PR gets integrated.

**Comment by fg1417** [comment]
Hi @nick-arm , can I have your review please?

**Review by nick-arm — APPROVED** [review]
Looks OK to me.

**Comment by fg1417** [comment]
> Looks OK to me.

Thanks, @nick-arm

**Pull request closed without merging**
